R/utils.R

```diff
@@ -154,6 +154,53 @@ random_name <- function(prefix = "") {
   paste0(prefix, "odbc_", name)
 }
 
+# error handling ---------------------------------------------------------------
+rethrow_database_error <- function(msg, call = trace_back()$call[[1]]) {
```

Comment:
The current draft of `rethrow_database_error()` tries to keep as much of the plain text of the original error message intact and in order as possible, just adding formatting, so that errors are still internet searchable. 

Another approach would just forefront simplicity/readibility: start with the vendor error message, clear out the "nanodbc/nanodbc<line#>" text (as it's probably only useful to odbc developers), and only include the error code when it's not nanodbc's 00000. A previous draft of this function did that:

```
Error in `dbConnect()`:
! Data source name not found and no default driver specified
ℹ From [unixODBC][Driver Manager].
```

Reply:
I wonder if we could split the difference and put the keep the nanodbc line number, but move it to the bottom. i.e. instead of:

```
i At 'nanodbc/nanodbc.cpp:1147', error 00000 from [unixODBC][Driver Manager].
x Data source name not found and no default driver specified
```

Do

```
i Error 00000 from [unixODBC][Driver Manager].
x Data source name not found and no default driver specified
i From 'nanodbc/nanodbc.cpp:1147'
```

And I think it's probably worth being explicit that this is an ODBC error, and it's unusual to start with an info bullet:

```
! ODBC failed with error 00000 from [unixODBC][Driver Manager].
x Data source name not found and no default driver specified
i From 'nanodbc/nanodbc.cpp:1147'
```


Also that's an unusual specification for the `call` argument.

Reply:
Totally on board. Just pushed some changes reflecting your last suggestion in [ebaf6dd](https://github.com/r-dbi/odbc/pull/789/commits/ebaf6dd9acbb94b95b37d94384392d023536c475), though am game to remove that last `i From 'nanodbc/nanodbc.cpp:1147'` line entirely if you are.

re: `call`, this is indeed wonky :/ `parent.frame()` (and thus `caller_env()`) don't seem to know how to traverse across the Rcpp code (since `rethrow_database_error()` is called from C++)—setting `call = caller_env(n)` with n of 0/1/2/3 gives no error context. 

There are certainly other ways of going about this (searching for a `call` in the environments listed by `trace_back()` or attempting to plumb a call through nanodbc.cpp)—let me know if you think it's worth spending more time there.

Reply:
Errors look good to me, and I think you just need to document why you need the weird `call` spec.